### PR TITLE
Festo controller

### PR DESCRIPTION
**Pull request by jgreefmiso**
This is a big boy.

The main function that you should care about if in controller_main.cpp. That contains a ROS wrapper for the actual controller code located in festo_cmmt_interface.cpp.

state_machine.cpp contains code to mirror the CiA402 state machine that is running on the Festo unit. Decent documentation for that state machine (which is general for EtherCAT products) is at https://en.nanotec.com/products/manual/PD2C_CAN_EN/general%252Fds402_power_state_machine.html/
Most of the states that actually do things are sub-states of Operation Enabled. Those sub-states represent Festo specific modes of operation, which are documented in the Festo manual https://www.festo.com/net/SupportPortal/Files/505161/CMMT-AS-SW_2018-06_8093861g1.pdf

### CODE REVIEW

File: festo_ethercat/include/festo_ethercat/festo_cmmt_interface.h [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+   * @return True if the rail is at position, false otherwise.
+   */
+
+  bool atPosition(int32_t);
+  /**
+   * Get the current rail position.
+   * @return The rail's position in millimeters
+   */
+
+  int32_t getPostion(void);
```
Comment (ludwig) [inline festo_ethercat/include/festo_ethercat/festo_cmmt_interface.h]: Typo in `getPosition()`?

Comment (ludwig) [reply festo_ethercat/include/festo_ethercat/festo_cmmt_interface.h]: I'm guessing you used tab completion a lot 😆

Comment (jgreefmiso) [reply festo_ethercat/include/festo_ethercat/festo_cmmt_interface.h]: Fixed.

**Comment by Schwo0ps** [comment]
@jgreefmiso Do you have automated tests? Could we set them up before merging this. I'll be a lot more confident about reviewing the PR if I can read the tests and see how they pass.

**Comment by jgreefmiso** [comment]
> @jgreefmiso Do you have automated tests? Could we set them up before merging this. I'll be a lot more confident about reviewing the PR if I can read the tests and see how they pass.

We don't have a test bed for this, so there's not much I can actually test. I could do something like "tell it to move and see if it self reports that it moved" but there'll be no external validation.

At a high level, what tests were you thinking?

File: festo_ethercat/src/festo_ethercat/controller_main.cpp [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+{
+  festo_interface_.setSpeedOverride(msg->speed_override);
+}
+
+void RailController::publishPositionCallback(const ros::TimerEvent &event)
+{
+  double rail_pos_m = ((double)festo_interface_.getPosition()) / 1000.0;
+  double rail_vel_m_s = ((double)festo_interface_.getVelocity()) / 1000.0;
+  sensor_msgs::JointState msg;
+  msg.header.seq = joint_state_msg_number_;
```
Comment (zzv2) [inline festo_ethercat/src/festo_ethercat/controller_main.cpp]: I don't think you are actually supposed to set this field:
https://answers.ros.org/question/55126/why-does-ros-overwrite-my-sequence-number/
Then you can remove the `joint_state_msg_number_` member var.

Comment (jgreefmiso) [reply festo_ethercat/src/festo_ethercat/controller_main.cpp]: Fixed

File: festo_ethercat/scripts/remove_sudo_req.sh [hunk 3]
```diff
@@ -0,0 +1,4 @@
+
+#!/bin/bash
```
Comment (ludwig) [inline festo_ethercat/scripts/remove_sudo_req.sh]: The `#!` should always be on the first line:
* https://stackoverflow.com/questions/12910744/why-should-the-shebang-line-always-be-the-first-line

Comment (jgreefmiso) [reply festo_ethercat/scripts/remove_sudo_req.sh]: Fixed.

File: festo_ethercat_msgs/srv/SetVelocity.srv [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# Set the veloctity at which the rail tries to move in meters per second.
+float64 velocity_m_s
```
Comment (rsinnet) [inline festo_ethercat_msgs/srv/SetVelocity.srv]: All velocities in the system should be in m/s so I guess no need to name the variable this way?

File: festo_ethercat/test/test_state_machine.cpp [hunk 5]
```diff
@@ -0,0 +1,525 @@
+
```
Comment (rsinnet) [inline festo_ethercat/test/test_state_machine.cpp]: More testing than I was expecting. Nice. I don't think this is being linted.

File: festo_ethercat/scripts/cmmt_controller.py [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+from collections import namedtuple
+import struct
+import sys
+import threading
+import time
+
+import pysoem
+from state_machine import Commands, InitialState
+
+class CmmtController:
```
[the remark this answers is not in the document]
Comment (jgreefmiso) [reply festo_ethercat/scripts/cmmt_controller.py]: I didn't expect this file to really be looked at, so I didn't prepare it for review, fixed now.

File: festo_ethercat_msgs/srv/SetVelocity.srv (hunk 4 above)
Comment (jgreefmiso) [reply festo_ethercat_msgs/srv/SetVelocity.srv]: The rail is in mm, so I'm going to keep this to make it clear in the rail code.

File: festo_ethercat/test/test_state_machine.cpp (hunk 5 above)
Comment (jgreefmiso) [reply festo_ethercat/test/test_state_machine.cpp]: Correct. I ran the clang formater (is that what you mean by linter?) and it made a lot of my code impossible to read, so I'm not using it.

**Pull request merged**